spring-integration-hazelcast/src/main/java/org/springframework/integration/hazelcast/outbound/HazelcastCacheWritingMessageHandler.java

```diff
@@ -16,64 +16,147 @@
+	private void writeToCache(final Message<?> message) {
+		DistributedObject distributedObject = getDistributedObject(message);
+		if (distributedObject instanceof IMap) {
+			((IMap) distributedObject).put(parseKeyExpression(message),
+					getPayloadOrMessage(message));
+		}
+		else if (distributedObject instanceof MultiMap) {
+			((MultiMap) distributedObject).put(parseKeyExpression(message),
+					getPayloadOrMessage(message));
+		}
+		else if (distributedObject instanceof ReplicatedMap) {
+			((ReplicatedMap) distributedObject).put(parseKeyExpression(message),
+					getPayloadOrMessage(message));
+		}
+		else if (distributedObject instanceof IList) {
+			((IList) distributedObject).add(getPayloadOrMessage(message));
+		}
+		else if (distributedObject instanceof ISet) {
+			((ISet) distributedObject).add(getPayloadOrMessage(message));
+		}
+		else if (distributedObject instanceof IQueue) {
+			((IQueue) distributedObject).add(getPayloadOrMessage(message));
 		}
-		else if (this.distributedObject instanceof IList) {
-			((IList<?>) this.distributedObject).addAll((List) message.getPayload());
+		else if (distributedObject instanceof ITopic) {
+			((ITopic) distributedObject).publish(getPayloadOrMessage(message));
 		}
-		else if (this.distributedObject instanceof ISet) {
-			((ISet<?>) this.distributedObject).addAll((Set) message.getPayload());
+	}
+
+	private DistributedObject getDistributedObject(final Message<?> message) {
+		if (this.distributedObject != null) {
+			return this.distributedObject;
+		}
+		else if (StringUtils.hasText(this.cacheExpression)) {
+			return parseCacheExpression(message);
+		}
+		else if (message.getHeaders().get(HazelcastHeaders.CACHE_NAME) != null) {
+			return this.getBeanFactory().getBean(
+					message.getHeaders().get(HazelcastHeaders.CACHE_NAME).toString(),
+					DistributedObject.class);
+		}
+		else {
+			throw new IllegalStateException("One of 'cache', 'cache-expression' and "
+					+ HazelcastHeaders.CACHE_NAME
+					+ " must be set for cache object definition.");
+		}
+	}
+
+	private DistributedObject parseCacheExpression(final Message<?> message) {
+		Object cacheName = parseExpression(this.cacheExpression, message);
+		return this.getBeanFactory().getBean(cacheName.toString(),
+				DistributedObject.class);
+	}
+
+	private Object parseKeyExpression(final Message<?> message) {
+		if (StringUtils.hasText(this.keyExpression)) {
+			return parseExpression(this.keyExpression, message);
+		}
+		else {
+			throw new IllegalStateException(
+					"'key-expression' must be set for IMap, MultiMap and ReplicatedMap");
+		}
+	}
+
+	private Object parseExpression(final String expressionString, final Message<?> message) {
+		Expression expression = new SpelExpressionParser()
+				.parseExpression(expressionString);
+		return expression.getValue(new StandardEvaluationContext(message));
+	}
+
+	private Object getPayloadOrMessage(final Message<?> message) {
```

Comment:
Not sure in this method. We can just extract `objectToStore` variable in the beginning of `handleMessageInternal`  and do the rest of logic.